contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
 	PG_RETURN_NUMERIC(result);
 }
 
+Datum
+parsename(PG_FUNCTION_ARGS)
+{
+	text *object_name = PG_GETARG_TEXT_PP(0);
+	int object_piece = PG_GETARG_INT32(1);
+	char *object_name_str = text_to_cstring(object_name);
+	UErrorCode errorCode = U_ZERO_ERROR;
+	UChar unicode_object_name_str[256] = {0};
+	int len;
+	UChar requested_object_part[256] = {0};
+	int current_part = 1;
+	int state = 0; // 0: outside quotes and brackets, 1: inside quotes, 2: inside brackets
+	int improperly_enclosed = 0;
+	int max_parts = 1;
+	int max_length;
+	int part_lengths[256] = {0};
+	
+	// Calculate max_parts based on the number of dots in the input string
+	for (char *dot = strchr(object_name_str, '.'); dot != NULL; dot = strchr(dot + 1, '.'))
+	{
+		max_parts++;
+	}
+    // Convert the input string to Unicode using ICU
+	u_strFromUTF8(unicode_object_name_str, sizeof(unicode_object_name_str) / sizeof(UChar), NULL, object_name_str, -1, &errorCode);
+	if (U_FAILURE(errorCode))
+	{
+		pfree(object_name_str);
+		PG_RETURN_NULL();
+	}
+	len = u_strlen(unicode_object_name_str);
+	if (object_piece < 1 || object_piece > max_parts)
+	{
+		PG_RETURN_NULL();
+	}
+	// Change the requested object_piece to reverse the order
+	object_piece = max_parts - object_piece + 1;
+	// Determine the maximum allowed length based on whether the input is Unicode or not
+	max_length = (len != strlen(object_name_str)) ? 255 : 128;
+	// Parsing loop to calculate part_lengths
+	for (int i = 0; i < len; i++)
+	{
+		UChar c = unicode_object_name_str[i];
+		if (state == 0) // Outside quotes and brackets
+		{
+			if (c == '"')
+			{
+				state = 1;
```

Comment:
I think you need to distinguish 2 cases in your state machine 
abc."abcd"  and 
abc.ab"cd" 
The second one is a syntax error